state.go

```diff
@@ -157,92 +235,67 @@ func (s *State) pipe(data, t string, command []string) (string, error) {
 		os.Environ(),
 		"CLAWS_PIPE_TYPE="+t,
 		"CLAWS_SESSION="+strconv.FormatInt(sessionStarted.UnixNano()/1000, 10),
-		"CLAWS_CONNECTION="+strconv.FormatInt(connectionStarted.UnixNano()/1000, 10),
-		"CLAWS_WS_URL="+s.Conn.URL(),
+		"CLAWS_CONNECTION="+strconv.FormatInt(s.ConnectionStarted.UnixNano()/1000, 10),
+		"CLAWS_WS_URL="+s.wsConn.URL(),
 	)
 	// set up stdin
-	stdin := strings.NewReader(data)
+	stdin := bytes.NewReader(data)
 	c.Stdin = stdin
 
 	// run the command
-	res, err := c.Output()
-	return string(res), err
+	return c.Output()
 }
 
-func (s *State) printToOut(f func(io.Writer, ...interface{}) (int, error), str string) {
-	if s.Settings.Timestamp != "" {
-		str = time.Now().Format(s.Settings.Timestamp) + str
+func (s *State) printToOut(
+	str string,
+	bIndent bool,
+	f func(io.Writer, ...interface{}) (int, error),
+) {
+
+	// NOTE: mutexed to sequentialize whole writes between
+	//       UI goroutine, read pump, & write pump
+	s.writerLock.Lock()
+	defer s.writerLock.Unlock()
+
+	var szTs string
+	oSet := s.Settings.Clone()
+	if len(oSet.Timestamp) > 0 {
+		szTs = time.Now().Format(oSet.Timestamp)
 	}
 
-	if str != "" && str[len(str)-1] != '\n' {
-		str += "\n"
-	}
 	s.ExecuteFunc(func(*gocui.Gui) error {
-		_, err := f(s.Writer, str)
-		return err
-	})
-}
-
-// Settings contains persistent information about the usage of claws.
-type Settings struct {
-	Info             string
-	JSONFormatting   bool
-	Timestamp        string
-	LastWebsocketURL string
-	LastActions      []string
-	Pipe             struct {
-		In  []string
-		Out []string
-	}
-}
 
-// Load loads settings from ~/.config/claws.json
-func (s *Settings) Load() error {
-	folder, err := getConfigFolder()
-	if err != nil {
-		return err
-	}
-
-	f, err := os.Open(folder + "claws.json")
-	if err != nil {
-		// silently ignore ErrNotExist
-		if os.IsNotExist(err) {
-			return nil
+		// TIMESTAMP, NOT INDENTED
+		bHasTs := len(szTs) > 0
+		if bHasTs {
+			if _, e := f(s.Writer, szTs); e != nil {
+				return e
+			}
 		}
-		return err
-	}
-	defer f.Close()
 
-	return json.NewDecoder(f).Decode(s)
-}
+		if bIndent {
 
-// Save saves settings to ~/.config/claws.json
-func (s Settings) Save() error {
-	folder, err := getConfigFolder()
-	if err != nil {
-		return err
-	}
+			if bHasTs {
+				if _, e := f(s.Writer, "\n"); e != nil {
+					return e
+				}
+			}
 
-	f, err := os.Create(folder + "claws.json")
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-
-	s.Info = "Claws configuration file; more information can be found at https://howl.moe/claws"
-	e := json.NewEncoder(f)
-	e.SetIndent("", "\t")
-	return e.Encode(s)
-}
+			const INDENT = "  "
+			sLines := strings.Split(str, "\n")
+			for _, szLine := range sLines {
+				if _, e := f(s.Writer, INDENT+szLine+"\n"); e != nil {
+					return e
+				}
```

Comment:
`f(s.Writer, indentPrefix + strings.Replace(str, "\n", "\n" + indentPrefix) + "\n")`?

Reply:
broke it up into a loop while experimenting.  will change back to ReplaceAll on the next commit.